orca-core/src/main/java/com/netflix/spinnaker/orca/locks/LockManager.java

```diff
@@ -0,0 +1,121 @@
+ *
+ *
+ * lockValue = new LockValue(application, execution.type, execution.id)
+ * createServerGroup
+ *   acquireLock(clusterName, lockValue, createServerGroup.stage.id)
+ *   deploy
+ *   waitForUpInstances
+ *
+ *   disableCluster
+ *     acquireLock(clusterName, lockValue, disableCluster.stage.id)
+ *     disableCluster
+ *     waitForDisableCluster
+ *     releaseLock(clusterName, lockValue, disableCluster.stage.id)
+ *
+ *   scaleDownCluster
+ *     acquireLock(clusterName, lockValue, scaleDownCluster.stage.id)
+ *     scaleDownCluster
+ *     waitForScaleDown
+ *     releaseLock(clusterName, lockValue, scaleDownCluster.stage.id)
+ *
+ *   releaseLock(clusterName, lockValue, createServerGroup.stage.id)
+ *
+ *
+ * The lifespan of the lock would be the entire duration of the createServerGroup stage including all
+ * the child stages (disableCluster, scaleDownCluster).
+ */
+public interface LockManager {
+
+  class LockValue {
+    final String application;
+    final String type;
+    final String id;
+
+    public LockValue(String application, String type, String id) {
+      this.application = Objects.requireNonNull(application);
+      this.type = Objects.requireNonNull(type);
+      this.id = Objects.requireNonNull(id);
+    }
+
+    public String getApplication() {
+      return application;
+    }
+
+    public String getType() {
+      return type;
+    }
+
+    public String getId() {
+      return id;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      LockValue lockValue = (LockValue) o;
+      return Objects.equals(application, lockValue.application) &&
+        Objects.equals(type, lockValue.type) &&
+        Objects.equals(id, lockValue.id);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(application, type, id);
+    }
+  }
+
+  /**
+   * Acquires a named lock.
+   * @param lockName The name of the lock.
+   * @param lockValue The value of the lock - if the lock is already held with this value, acquisition is successful otherwise lock acquisition fails.
+   * @param lockHolder The holder of the lock.
+   * @param ttlSeconds How long to acquire for or extend an existing lock for.
+   * @throws LockFailureException if the lock is currently held with a different lockValueApplication/lockValueType/lockValue
+   */
+  void acquireLock(String lockName, LockValue lockValue, String lockHolder, int ttlSeconds) throws LockFailureException;
+
+  /**
+   * Extends a named lock ttl.
+   * @param lockName The name of the lock.
+   * @param lockValue The value of the lock - if the lock is already held with this value, the ttl extension is successful otherwise lock extension fails.
+   * @param ttlSeconds How long to extend the lock for.
+   * @throws LockFailureException if the lock is currently held with a different lockValueApplication/lockValueType/lockValue
```

Comment:
what's the expected behavior if you attempt to extend a lock that does not exist (e.g. expired or it slipped on a banana on its way to redis)